### PR TITLE
feat(frontend): BtcConvertTokenWizard component

### PR DESCRIPTION
**Pull request by DenysKarmazynDFINITY**
# Motivation

A wizard component for controlling the conversion flow.

### CODE REVIEW

File: src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	import type { NetworkId } from '$lib/types/network';
+	import type { OptionAmount } from '$lib/types/send';
+	import { invalidAmount, isNullishOrEmpty } from '$lib/utils/input.utils';
+	import {
+		isNetworkIdBTCRegtest,
+		isNetworkIdBTCTestnet,
+		mapNetworkIdToBitcoinNetwork
+	} from '$lib/utils/network.utils';
+
+	export let currentStep: WizardStep | undefined;
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte:39]: There will be 2 more parents: ConvertModal (currentStep is defined here) -> ConvertWizard -> BtcConvertWizard (we are here).

File: src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	let sourceAddress: string;
+	$: sourceAddress =
+		(isNetworkIdBTCTestnet(networkId)
+			? $btcAddressTestnet
+			: isNetworkIdBTCRegtest(networkId)
+				? $btcAddressRegtest
+				: $btcAddressMainnet) ?? '';
+
+	let destinationAddress: string | undefined = undefined;
+	$: destinationAddress = $btcAddressStore?.[$destinationToken.id]?.data;
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte:67]: Before opening a BTC -> ckBTC modal, I'm checking if ckBTC info has been loaded or not (`loadAllCkBtcInfo` method) - it makes sure that we have `$btcAddressStore` loaded for the current `$destinationToken.id` (in our case - ckBTC token id), as well as `kyt_fee`. The same pattern is used in the existing ckBTC -> BTC conversion flow.

File: src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+
+		if (isNullish($authIdentity)) {
+			await nullishSignOut();
+			return;
+		}
+
+		// The data has been already validated in the previous conversion flow steps.
+		if (
+			isNullish(network) ||
+			isNullishOrEmpty(destinationAddress) ||
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte:82]: Yes, it is checked before opening the modal (ConvertoCkBTC component, not merged yet).

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
+import BtcConvertTokenWizard from '$btc/components/convert/BtcConvertTokenWizard.svelte';
+import * as btcSendServices from '$btc/services/btc-send.services';
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts:2]: I would say a bit complicated - the number of mocks would only grow if I do so (e.g. instead of mocking sendBtc service, I would need to do so for sendBtcApi, addPendingBtcTransaction, and probably waitAndTriggerWallet). In any case, we won't be able to get the "real" result of a service call (which ideally is unit-tested separately), so I would probably avoid the complexity.

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+import * as utxosFeeStore from '$btc/stores/utxos-fee.store';
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts:3]: Yes, this component initialises this store, hence doesn't access the value via context. Therefore, we need to mock the utxos value if we want to pass validations.

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+import type { Token } from '$lib/types/token';
+import { mockBtcAddress, mockUtxosFee } from '$tests/mocks/btc.mock';
+import { mockIdentity } from '$tests/mocks/identity.mock';
+import { mockPage } from '$tests/mocks/page.store.mock';
+import type { Identity } from '@dfinity/agent';
+import { assertNonNullish } from '@dfinity/utils';
+import { fireEvent, render } from '@testing-library/svelte';
+import { readable } from 'svelte/store';
+
+describe('BtcConvertTokenWizard', () => {
```
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts:23]: I added 3 more tests to make sure that we render a proper step component. Is this something you had in mind?

**Comment by DenysKarmazynDFINITY** [comment]
@peterpeterparker ready for another look.

File: src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte:67]: Right, thanks for the reminder. So the destination address is the BTC address generated by the ckBTC minter to which the user needs to send BTC if they want to wrap BTC -> ckBTC, correct?

If yes, worth a small comment as description as reminder or document it somewhere (if not already the case)?

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts:2]: I'm hesitant. I definitely understand your point. On the other, I'm a bit afraid that the complexity and coverage grows but, we actually miss some edge case because we do mock at a too high level. How complicated it will be to setup E2E tests for this conversion? If we will be able to have E2E test, I would say it's fine to mock the service. If we will never have E2E test for it, I would say it would be better to not mock the service. WDYT?

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts:23]: I was more thinking at testing the all flow but, this goes a bit in the same direction as my previous question about the vision of the E2E tests but, given that this component will still be consumed by parents, I guess it's for a later point to ask.

File: src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte (hunk 2 above)
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/btc/components/convert/BtcConvertTokenWizard.svelte:69]: Correct! Added a comment with the link to the docs.

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (DenysKarmazynDFINITY) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts]: Totally makes sense. It will definitely take some time until we have e2e tests here, could be quite tricky to somehow mock different APIs to be able to use this and all other flows "natively". So, for now, let's indeed move one layer down and do not touch the service in the unit tests. I am now mocking the underlying APIs. Please take a look.

**Review by peterpeterparker — APPROVED** [review]
LGTm, thx

File: src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (peterpeterparker) [reply src/frontend/src/tests/btc/components/convert/BtcConvertTokenWizard.spec.ts]: Thanks!
